File: mooc/binTree.c
//
//  Tree.c
//  cCode
//
//  Created by clockworkSong on 2020/2/9.
//  Copyright © 2020 clockworkSong. All rights reserved.
//

#include <stdio.h>
#include <stdlib.h>

typedef struct BinNode{
    int data;
    struct BinNode* left;
    struct BinNode* right;
}*pBinTree;

pBinTree initBinTree(int x){
    pBinTree b;
    b = (pBinTree)malloc(sizeof(struct BinNode));
    b->data = x;
    b->left = b->right = NULL;
    return b;
}

//前序遍历
void preOrder(const pBinTree p)
{
    if(p == NULL) return;
    printf("%d ", p->data);
    preOrder(p->left);
    preOrder(p->right);
}

pBinTree findX(pBinTree r, int x){
    if(r == NULL) return NULL;
    if(r->data == x) return r;
    pBinTree f = findX(r->left, x);
    return f ? f : findX(r->right, x);
}

int insertBinTree(pBinTree pb, int rdata, int lr, int data){
    pBinTree f = findX(pb, rdata);
    if(f == NULL) return 0;
    if(lr == 0){
        if(f->left != NULL) return 0;
        f->left = initBinTree(data);
    }
    else{
        if(f->right != NULL) return 0;
        f->right = initBinTree(data);
    }
    return 1;
}

int main(int argc, const char * argv[]) {
    pBinTree t = initBinTree(11);
    insertBinTree(t, 11, 0, 22);
    insertBinTree(t, 11, 1, 33);
    insertBinTree(t, 22, 0, 44);
    insertBinTree(t, 33, 0, 55);
    preOrder(t);
    printf("\n");

    return 0;
}

File: mooc/binTree.py

#二叉树节点
class BinNode:
	def __init__(self, data):# init构造函数
		self.data = data
		self.left = None
		self.right = None
#表示完整的二叉树
class BinTree(object):
	"""docstring for BinTree"""
	def __init__(self, arg):
		super(BinTree, self).__init__()
		self.root = BinNode(arg)# 将树根初始化为传入的arg

	def preOrder(self, r):
		if r is None:
			return
		print(r.data, end=' ')
		self.preOrder(r.left)
		self.preOrder(r.right)

	def findX(self, pb, x):
		if pb is None:
			return
		if pb.data == x:
			return pb
		f = self.findX(pb.left, x)
		if f is not None:
			return f
		return self.findX(pb.right, x)

	def insert(self, pb, rdata, lr, data):
		f = self.findX(pb, rdata)
		if f is None:
			return
		if lr == 0:
			if f.left is not None:
				return
			f.left = BinNode(data)
		else:
			if f.right is not None:
				return
			f.right = BinNode(data)

a = BinTree(11)
a.insert(a.root, 11, 0, 22)
a.insert(a.root, 11, 1, 33)
a.insert(a.root, 22, 1, 44)
a.insert(a.root, 22, 0, 55)
a.insert(a.root, 33, 1, 66)
a.preOrder(a.root)


